Bio/SeqUtils/__init__.py

```diff
@@ -667,6 +668,42 @@ def calculate(self, sequence):
 
         return exp(cai_value / cai_length)
 
+    def optimize(self, sequence, seq_type="DNA"):
+        """Return a new DNA sequence with preferred codons only."""
+        try:  # If seq record is provided, convert to sequence
+            sequence = sequence.seq
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {}
+        for codon, aminoacid in self._table.forward_table.items():
+            if self[codon] == 1.0:
+                if aminoacid in pref_codons:
+                    message = f"{pref_codons[aminoacid]} and {codon} are equally preferred. Using {codon}"
+                    warnings.warn(message, RuntimeWarning)
+                pref_codons[aminoacid] = codon
+        for codon in self._table.stop_codons:
+            if self[codon] == 1.0:
+                pref_codons["*"] = codon
+        # Create amino acid sequence if DNA was provided
+        if seq_type == "DNA" or seq_type == "RNA":
+            aa_seq = translate(seq)
+        elif seq_type == "protein":
+            aa_seq = seq
+        else:
+            raise TypeError
+        # Un-translate in loop using only preferred codons
+        optimized = []
+        for aa in aa_seq:
+            try:
+                pref_codon = pref_codons[aa]
+            except KeyError:
+                print(f"Unrecognized amino acid: {aa}")
+            optimized.append(pref_codon)
```

Comment:
Two problems:
- first letter is unrecognized -> `UnboundLocalError`
- second letter is unrecognized -> repeats first letter

If you intent to drop those unrecognized amino acids, I suggest to add a test like this:
```py
assert cai.optimize("EDU", "protein").translate() == "ED"
assert cai.optimize("EUD", "protein").translate() == "ED"
assert cai.optimize("UED", "protein").translate() == "ED"
```

Reply:
Raising an exception (ValueError? Maybe KeyError) would seem fine too - and you can write tests to confrim it raises an exception.

Reply:
Thanks, good catch. I've made changes to raise a KeyError on invalid codons and added a test to confirm a KeyError is raised.

Reply:
Looks good to me now, thanks for the changes.

FYI, I would probably opt for this :-)
```py
        try:
            optimized_str = "".join(pref_codons[aa] for aa in seq)
        except KeyError as ex:
            raise KeyError(f"Unrecognized amino acid: {ex}") from None
```

Reply:
That looks shorter/faster, but is the exception as helpful? i.e. Does it tell you which amino acid was a problem

Reply:
Yes, the exception will be the same, `str(ex)` will be the unrecognized amino acid.